Make layout supports compatible with enhanced pagination

luisherranz opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

This is a backport PR for WordPress 6.4 that includes the following PHP Gutenberg changes:

- https://github.com/WordPress/gutenberg/pull/55416

Trac ticket: https://core.trac.wordpress.org/ticket/59681

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

luisherranz committed:
Add the prefix-based unique id function

luisherranz committed:
Use it to create the class names

tellthemachines committed:
Fix failing render test

tellthemachines reviewed (commented):
Thanks for the PR! The code change itself LGTM; my question is whether the layout file is the best home for the `wp_incremental_id_per_prefix` function. Might we be using it elsewhere in the future? 

Also might be good to have a unit test for the function, similar to `wp_unique_id` [here](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/functions.php#L1313).

I pushed a fix for the failing PHP tests that were still expecting the old classnames.

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -540,6 +540,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
+ * @param  string $prefix Prefix for the returned ID.
+ * @return string         Incremental ID per prefix.
+ */
+function wp_incremental_id_per_prefix( $prefix = '' ) {
```
Regarding what @tellthemachines said about defining the function in this file, I think this function should be placed in:

- this file, if it is only intended to be used for blocks and general usage is not really desired. However, if this is the case, I'd suggest renaming the function so that its intended limited context is clear. `wp_unique_block_layout_classname()`, for example.
  - if this is the final location, tests can be added for this function in `tests/phpunit/tests/blocks/wpUniqueBlockLayoutClassname.php`, with the class name `Tests_Blocks_WpUniqueBlockLayoutClassname`
- in `wp-includes/functions.php`, just below `wp_unique_id()`, if general use is not discouraged. However, if this is the case, I think `wp_unique_prefixed_id()` may be a better name as it will align better with `wp_unique_id()`.
  - if this is the final location, tests can be added for this function in `tests/phpunit/tests/functions/wpUniquePrefixedId.php` with the class name `Tests_Functions_WpUniquePrefixedId.php`.

costdev replied:
Fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09). 👍

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -540,6 +540,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
+/**
+ * Generates an incremental ID that is independent per each different prefix.
+ *
+ * It is similar to `wp_unique_id`, but each prefix has it's own internal ID
```
```suggestion
 * It is similar to `wp_unique_id`, but each prefix has its own internal ID
```

costdev replied:
Fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09). 👍

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -540,6 +540,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
+ * prefix.
+ *
+ * @param  string $prefix Prefix for the returned ID.
+ * @return string         Incremental ID per prefix.
```
```suggestion
 * @return string Incremental ID per prefix.
```
The `@return` description should not be aligned with the `@param` descriptions. See [the Documentation standard](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods) for an example of this.

costdev replied:
Fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09). 👍

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -630,7 +650,16 @@ function wp_render_layout_support_flag( $block_content, $block ) {
+	*/
+	$container_class = wp_incremental_id_per_prefix(
+		'wp-container-' . sanitize_title( $block['blockName'] ) . '-layout-'
+	);
```
```suggestion
	/*
	 * We use an incremental ID that is independent per prefix to make sure that
	 * rendering different numbers of blocks doesn't affect the IDs of other
	 * blocks. We need this to make the CSS class names stable across paginations
	 * for features like the enhanced pagination of the Query block.
	 */
	$container_class = wp_incremental_id_per_prefix(
		'wp-container-' . sanitize_title( $block['blockName'] ) . '-layout-'
	);
```
Slight alignment issue.

costdev replied:
Partially fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09).

See [this comment](https://github.com/WordPress/wordpress-develop/pull/5528#discussion_r1367524573) for the remaining alignment issue.

costdev commented on tests/phpunit/tests/blocks/render.php:
```diff
@@ -218,9 +218,9 @@ public function test_do_block_output( $html_filename, $server_html_filename ) {
 		$html = do_blocks( self::strip_r( file_get_contents( $html_path ) ) );
 		// If blocks opt into Gutenberg's layout implementation
-		// the container will receive an added classname of `wp_unique_id( 'wp-container-' )`
+		// the container will receive an added classname of `wp_unique_id( 'wp-container-[blockname]-layout' )`
```
```suggestion
		// the container will receive an added classname of `wp_unique_id( 'wp-container-[blockname]-layout' )`
```
This function name is no longer relevant here. To save reference to implementation, I'd suggest this instead:

```suggestion
		// the container will receive an additional, unique classname based on "wp-container-[blockname]-layout"
```

Additionally, the next line mentions an `id`, but it's actually referring to a classname, so this could be changed for accuracy while we're here.

costdev replied:
Partially fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09). The `id` on the next line wasn't changed, but this isn't a critical issue.

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -540,6 +540,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
+ * @param  string $prefix Prefix for the returned ID.
+ * @return string         Incremental ID per prefix.
+ */
+function wp_incremental_id_per_prefix( $prefix = '' ) {
```
Does an empty string prefix make sense here?

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -540,6 +540,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
+ * but it is unique across the life of the PHP process and it's stable per
+ * prefix.
+ *
+ * @param  string $prefix Prefix for the returned ID.
```
This optional parameter should be documented as such:
```suggestion
 * @param  string $prefix Optional. Prefix for the returned ID. Default empty string.
```

costdev replied:
Fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09). 👍

costdev reviewed (commented):
Thanks for the PR @luisherranz and for the ping @tellthemachines!

I've left some thoughts in this review.

mukeshpanchal27 commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -540,6 +540,26 @@ function wp_get_layout_style( $selector, $layout, $has_block_gap_support = false
+ */
+function wp_incremental_id_per_prefix( $prefix = '' ) {
+	static $id_counters = array();
+	if ( ! array_key_exists( $prefix, $id_counters ) ) {
+			$id_counters[ $prefix ] = 0;
+	}
```
```suggestion
	if ( ! array_key_exists( $prefix, $id_counters ) ) {
		$id_counters[ $prefix ] = 0;
	}
```

Indention.

costdev replied:
Fixed in [2d3db44](https://github.com/WordPress/wordpress-develop/pull/5528/commits/2d3db4418b82b8860fd10109cdb9fefaa4b2be09). 👍

mukeshpanchal27 commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -630,7 +650,16 @@ function wp_render_layout_support_flag( $block_content, $block ) {
 	$layout_definitions = wp_get_layout_definitions();
-	$container_class    = wp_unique_id( 'wp-container-' );
+
+	/*
+	* We use an incremental ID that is independent per prefix to make sure that
+	* rendering different numbers of blocks doesn't affect the IDs of other
+	* blocks. We need this to make the CSS class names stable across paginations
+	* for features like the enhanced pagination of the Query block.
+	*/
```
```suggestion
	/*
	 * We use an incremental ID that is independent per prefix to make sure that
	 * rendering different numbers of blocks doesn't affect the IDs of other
	 * blocks. We need this to make the CSS class names stable across paginations
	 * for features like the enhanced pagination of the Query block.
	 */
```

cbravobernal committed:
Add changes requested

cbravobernal commented:
Hi there!

I applied all changes requested. Thanks for the review @tellthemachines @costdev

costdev commented on tests/phpunit/tests/functions/wpUniquePrefixedId.php:
```diff
@@ -0,0 +1,26 @@
+		$second_default = wp_unique_prefixed_id();
+		$this->assertNotEquals( $first, $second );
+		$this->assertNotEquals( $default, $second_default );
+		$this->assertNotEquals( $null_id, $second_null_id );
```
If one of these assertions fails, the others will not run. Since there's multiple `$prefix` values being tested here, we can use a data provider to ensure every test runs even if a previous one fails.

We can also test the values to make sure behaviour is protected going forward.

Here's how that looks:
```php
/**
 * Tests that the expected unique prefixed IDs are created.
 *
 * @ticket 59681
 *
 * @dataProvider data_should_create_unique_prefixed_ids
 *
 * @param mixed $prefix   The prefix.
 * @param array $expected The next two expected IDs.
 */
public function test_should_create_unique_prefixed_ids( $prefix, $expected ) {
	$id1 = wp_unique_prefixed_id( $prefix );
	$id2 = wp_unique_prefixed_id( $prefix );

	$this->assertNotSame( $id1, $id2, 'The IDs are not unique.' );
	$this->assertSame( $expected, array( $id1, $id2 ), 'The IDs did not match the expected values.' );
}

/**
 * Data provider.
 *
 * @return array[]
 */
public function data_should_create_unique_prefixed_ids() {
	return array(
		'prefix as null'               => array(
			'prefix'   => null,
			'expected' => array( '1', '2' ),
		),
		'prefix as empty string'       => array(
			'prefix'   => '',
			'expected' => array( '3', '4' ),
		),
		'prefix as (string) "0"'       => array(
			'prefix'   => '0',
			'expected' => array( '01', '02' ),
		),
		'prefix as (int) 0'            => array(
			'prefix'   => 0,
			'expected' => array( '03', '04' ),
		),
		'prefix as string'             => array(
			'prefix'   => 'test',
			'expected' => array( 'test1', 'test2' ),
		),
		'prefix as string with spaces' => array(
			'prefix'   => '   ',
			'expected' => array( '   1', '   2' ),
		),
		'prefix as (string) "1"'       => array(
			'prefix'   => '1',
			'expected' => array( '11', '12' ),
		),
		'prefix as (int) 1'            => array(
			'prefix'   => 1,
			'expected' => array( '13', '14' ),
		),
	);
}
```

These are based on the current behaviour of the function. If any of these outcomes feels wrong, the source code should enforce the preferred behaviour.

costdev commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -630,7 +630,16 @@ function wp_render_layout_support_flag( $block_content, $block ) {
+	 * rendering different numbers of blocks doesn't affect the IDs of other
+	 * blocks. We need this to make the CSS class names stable across paginations
+	 * for features like the enhanced pagination of the Query block.
+	*/
```
```suggestion
	 */
```
Minor alignment issue.

costdev reviewed (changes requested):
Thanks for the updates @luisherranz!

I've left a note about a minor alignment issue, and an example of how we can leverage a data provider to ensure all tests run even if one fails.

cbravobernal committed:
Update tests, fix spacing in comment

cbravobernal commented:
Code updated with changes requested. Thanks for your reviews @tellthemachines , @costdev ! 🙇

luisherranz commented:
This looks much better now. Thanks, @c4rl0sbr4v0 and @costdev! 🙏

hellofromtonya commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -630,7 +630,16 @@ function wp_render_layout_support_flag( $block_content, $block ) {
-	$container_class    = wp_unique_id( 'wp-container-' );
+
+	/*
+	 * We use an incremental ID that is independent per prefix to make sure that
```
Avoid using "We" words in Core, as its not well understood who "we" refers to.

```suggestion
	 * Uses an incremental ID that is independent per prefix to make sure that
```

hellofromtonya commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -630,7 +630,16 @@ function wp_render_layout_support_flag( $block_content, $block ) {
+	/*
+	 * We use an incremental ID that is independent per prefix to make sure that
+	 * rendering different numbers of blocks doesn't affect the IDs of other
+	 * blocks. We need this to make the CSS class names stable across paginations
```
Avoid using "We" words in Core, as its not well understood who "we" refers to.
```suggestion
	 * blocks. Makes the CSS class names stable across paginations
```

the next line of text may require moving some of its words to this line of text, now that the string length is shortened.

cbravobernal committed:
Apply suggestions from code review

Co-authored-by: Tonya Mork <tonya.mork@automattic.com>

cbravobernal commented:
== Patch Report

Patch tested: https://patch-diff.githubusercontent.com/raw/WordPress/wordpress-develop/pull/5528.diff

How to read this report:
  🐞 <= Bug occurs.
  ✅ <= Behavior is ''expected''.
  ❌ <= Behavior is ''NOT expected''.

==== Environment
* OS: macOS 14.0
* Web Server: Nginx
- macOS 14.0
- PHP 8.2.11
- 6.4-RC1-56475-src
- Firefox 118.0.2

==== Steps to Reproduce
1. With WordPress 6.4 installed
2. Create 7 posts, configure to have 3 per page.
3. On the site editor, in home page, TT4, add pagination to the Query Loop, set it to inherit the template settings, activate "Enhanced Pagination feature). Make sure your layout is three posts in a row.
6. 🐞 Check that on page three, the last post layout is broken (full width instead of 1/3)

==== Steps to Test Patch
1. 🛠 Apply [https://patch-diff.githubusercontent.com/raw/WordPress/wordpress-develop/pull/5528.diff the patch].
2. Refresh the home page.
3. ✅ Check that the layout is correct when navigating to the last page.
4. ✅ Check that the navigation happened without a page reload.

==== Test Results
- ✅ Issue is reproducible.
- ✅ Issue resolved with patch.

==== Screenshots
Before patch:
![Screenshot 2023-10-23 at 17 58 31](https://github.com/WordPress/wordpress-develop/assets/37012961/2eb469a2-2b83-4808-8c2a-908f961648da)

After patch:
![Screenshot 2023-10-23 at 17 58 13](https://github.com/WordPress/wordpress-develop/assets/37012961/0814e120-2c74-4bbd-80c1-c143e220ea1f)

hellofromtonya commented on src/wp-includes/functions.php:
```diff
@@ -7830,6 +7830,28 @@ function wp_unique_id( $prefix = '' ) {
+ */
+function wp_unique_prefixed_id( $prefix = '' ) {
+	static $id_counters = array();
+	if ( ! array_key_exists( $prefix, $id_counters ) ) {
```
A performance effort has been underway to replace `array_key_exists()` with `isset()`, e.g. [Trac 57067](https://core.trac.wordpress.org/ticket/57067). 

To interchangeably use either `isset()` or `array_key_exists()`, the value cannot be `null`. 

In this new function, the value will always be an integer and never `null`.  See it in action here https://3v4l.org/1XvXO.

Therefore, `isset()` can be used as a teeny tiny micro-optimization and for consistency.

```suggestion
	if ( ! isset( $id_counters[ $prefix ] ) ) {
```

hellofromtonya commented on src/wp-includes/functions.php:
```diff
@@ -7830,6 +7830,28 @@ function wp_unique_id( $prefix = '' ) {
+	if ( ! array_key_exists( $prefix, $id_counters ) ) {
+		$id_counters[ $prefix ] = 0;
+	}
+	return $prefix . (string) ++$id_counters[ $prefix ];
```
An empty line before `return` for consistency.

```suggestion

	return $prefix . (string) ++$id_counters[ $prefix ];
```

hellofromtonya committed:
Use isset() instead of array_key_exists()

hellofromtonya commented on tests/phpunit/tests/functions/wpUniquePrefixedId.php:
```diff
@@ -0,0 +1,74 @@
+	 */
+	public function data_should_create_unique_prefixed_ids() {
+		return array(
+			'prefix as null'               => array(
```
The prefix should never be `null`,  as it's documented to require a `string`. And the return is also a `string`.

What happens if `null` is passed to it?
PHP automatically juggles the key to an empty string https://3v4l.org/NqAUJ.

Thinking though non-string values would not be part of the happy-path coverage and could be misleading or confusing to contributors reading the tests.

hellofromtonya replied:
Note: Separating the non-string values from the data provider is not blocking for the PR. The tests can be iterated upon after commit.

hellofromtonya replied:
cc @costdev

hellofromtonya commented on src/wp-includes/block-supports/layout.php:
```diff
@@ -630,7 +630,16 @@ function wp_render_layout_support_flag( $block_content, $block ) {
+	 * for features like the enhanced pagination of the Query block.
+	 */
+	$container_class = wp_unique_prefixed_id(
+		'wp-container-' . sanitize_title( $block['blockName'] ) . '-layout-'
```
Can the `blockName` be anything other than a `string` data type? I assume no, but am confirming after looking at the tests which pass non-string values as prefix to `wp_unique_prefixed_id()`.

aaronjorbin commented on src/wp-includes/functions.php:
```diff
@@ -7830,6 +7830,29 @@ function wp_unique_id( $prefix = '' ) {
+		$id_counters[ $prefix ] = 0;
+	}
+
+	return $prefix . (string) ++$id_counters[ $prefix ];
```
nit: I find the incrementing inside the return to be a bit harder to read. I wonder if it might be better to increment and then return.

hellofromtonya replied:
It's consistent with `wp_unique_id()`. That said, I thought the same thing : separate the increment from the string build/return.

cbravobernal replied:
Maybe we can refactor both in 6.5, as it is code quality. I can leave the PR ready after the RC, pointing trunk branch.

hellofromtonya replied:
I think what makes it less readable as compared to `wp_unique_id()` is the array lookup. To help, split the increment in 11f1cace5cb60369011f9e26e7ada5e412c70b57.

aaronjorbin commented on tests/phpunit/tests/functions/wpUniquePrefixedId.php:
```diff
@@ -0,0 +1,74 @@
+				'prefix'   => '1',
+				'expected' => array( '11', '12' ),
+			),
+			'prefix as (int) 1'            => array(
```
```suggestion
			'prefix as (int) 1 after (string) 1'            => array(
```

Unless the expectation is that `(int)` and `(string)1` should behave different (which I don't think it is), then I think it should be clearer why this starts at 13.

aaronjorbin commented on src/wp-includes/functions.php:
```diff
@@ -7830,6 +7830,29 @@ function wp_unique_id( $prefix = '' ) {
+ * @param string $prefix Optional. Prefix for the returned ID. Default empty string.
+ * @return string Incremental ID per prefix.
+ */
+function wp_unique_prefixed_id( $prefix = '' ) {
```
Since WP doesn't use typehints yet, I think it might be good to throw a `_doing_it_wrong` or some other warning if a nonstring is passed here since some of them can produce unexpected content.  See: https://3v4l.org/qb1k6

hellofromtonya replied:
@aaronjorbin and I chatted about this in Make/Core slack [starting here in the thread](https://wordpress.slack.com/archives/C02RQBWTW/p1698094191795039?thread_ts=1698066105.240609&cid=C02RQBWTW).

Consensus:

* `string` is expected, non-strings are not expected.
* When receiving a non-string, throw either a `wp_trigger_error()` or `_doing_it_wrong()` and then use the default of an empty string.

Following @azaozz concerns of overusing `_doing_it_wrong()`, i.e. should be used when it will break something and thus must bail out to avoid an error, I think it needs to use `wp_trigger_error()`. 

This change will affect the tests too. I'll prep follow-up commits to this PR to address these changes.

hellofromtonya replied:
Commit 42035dea6f875148d08094e2063d9515f17a61ff adds a defensive guard for a non-string per above -> raises a PHP notice via `wp_trigger_error()` and sets the `$prefix` to an empty string. Also splits the tests into happy and unhappy paths.

@aaronjorbin want to double check the changes?

aaronjorbin replied:
Looks Great. Thank you!

aaronjorbin commented on src/wp-includes/functions.php:
```diff
@@ -7830,6 +7830,29 @@ function wp_unique_id( $prefix = '' ) {
+ * but it is unique across the life of the PHP process and it's stable per
+ * prefix.
+ *
+ * @since 6.4
```
```suggestion
 * @since 6.4.0
```

hellofromtonya committed:
Defensive guard for nonstring

hellofromtonya committed:
Use full 6.4.0 version

hellofromtonya committed:
Increment separately and then return

hellofromtonya committed:
Run tests in separate processes.

The wp_unique_prefixed_id() function uses a function scoped static variable for memoization. The incremented value may be different between tests, especially over time when more tests are added.

Running the tests in separate process improves the stability of these tests.

hellofromtonya committed:
Fix wpcs issues

costdev commented on tests/phpunit/tests/functions/wpUniquePrefixedId.php:
```diff
@@ -0,0 +1,110 @@
+			'prefix as (string) "1"'       => array(
+				'prefix'   => '1',
+				'expected' => array( '11', '12' ),
+			),
```
As `wp_unique_prefixed_id()` is a general function not specific to block names, it's possible that an ID may contain `.` and a number of other characters. [Reference](https://www.w3.org/TR/html4/types.html#:~:text=ID%20and%20NAME%20tokens%20must,periods%20(%22.%22).).

It's also possible that the "id" in question does not link directly to the HTML `id` attribute`.

Coverage should also be added for a variety of possible "id" values, such as `(string) "0.0"`, etc. to protect _behavior_ as this general function could be used to generate unique IDs for a variety of purposes/accepted formats.

hellofromtonya replied:
Sure that can be added, though would be part of the prefix, and not the incremented integer value.

hellofromtonya replied:
Hey @costdev, commit 7a32e2600d86a513068a858e4289bc1d3a2ce04b adds 2 more happy path datasets:

* One for `'.'`
* Another for the prefix with a block name.

What do you think?

costdev replied:
Thanks Tonya!

If there are any other potential "gotcha" values that a potential future refactor could regress, we should consider those too.  The non-string dataset looks good for "gotcha" values, though I think we should complete the scalar datasets with `(float) 0.0` and `(float) 1.0`.

hellofromtonya replied:
@costdev 28cbcad9f9ceca4af4e6d784fa9b48d943a2b44f adds a double dataset.

hellofromtonya replied:
@costdev here's the added coverage we talked about in Make/Core slack DM 51514403078a9affd2ee03950a9814b356a8dc5f.

hellofromtonya committed:
Add additional happy path datasets

aaronjorbin commented on tests/phpunit/tests/functions/wpUniquePrefixedId.php:
```diff
@@ -0,0 +1,120 @@
+	 * @param string $expected_message  Expected notice message.
+	 * @param string $expected_value    Expected unique ID.
+	 */
+	public function test_should_raise_notice_and_use_empty_string_prefix_when_nonstring_given( $non_string_prefix, $expected_message, $expected_value ) {
```
I think this can be a post commit cleanup, but If this test is ever run before `test_should_create_unique_prefixed_ids` then they will fail.

hellofromtonya replied:
Yes, the order of the tests within the class does currently matter. The tests in this class runs in a separate process to avoid a situation where the static`$id_counters` already has values in it. 

To handle the test run order within the class, could change that to run each test in a separate process. By doing so, the static `$id_counters` will always be reset to an empty array when each test starts.

hellofromtonya replied:
Commit 5aac2049e9d55dd21d0c589b11108cc839fde67c runs each test dataset in a separate process to avoid test order concerns,

hellofromtonya replied:
I also tweaked the unhappy path tests to ensure the IDs remain unique f4f99fc242a565eda8548a678cdd8423faf13e65

How? By allowing each dataset to determine how many unique IDs to generate.

Why? Each of these non-string prefix datasets would have the same ID, as the prefix is omitted (empty string). One more check just to make sure.

hellofromtonya committed:
Tests: run each dataset separately.

To avoid test order, changes to run each dataset in a separate process, which each test starts with the static `$id_counters` equal to an empty array.

hellofromtonya committed:
Tests: Ensure unhappy path IDs are unique.

Unhappy path tests have non-string prefixes, which will be changed to the default empty string.

To further check uniqueness within the dataset's expected results, the unhappy path test now allows each dataset to specify how many unique IDs to generate for its tests.

hellofromtonya committed:
Test: Add double dataset

hellofromtonya committed:
Fix test dataset array => alignment

hellofromtonya commented:
@aaronjorbin @costdev all feedback is now addressed. Can you re-review please?

hellofromtonya committed:
Tests: Adds coverage for when given prefixes result in same prefix.

THis test guards against a future regression should the defensive guard ever change.

Props @costdev.

costdev reviewed (approved):
Thanks for the updates @luisherranz and @hellofromtonya! LGTM! 👍

hellofromtonya reviewed (approved):
LGTM 👍  should be ready for commit

ockham commented:
Thank you everyone! Luis has asked me to commit this, which I'll promptly do 😊

ockham commented:
Committed to Core `trunk` in [`[56994]`](https://core.trac.wordpress.org/changeset/56994) and to the `6.4` branch in [`[56995]`](https://core.trac.wordpress.org/changeset/56995).

luisherranz commented:
Thank you all for your help getting this to the finish line! @ockham, @c4rl0sbr4v0, @hellofromtonya, @aaronjorbin and @costdev 🙏
